floats/floats.go

```diff
@@ -512,34 +521,117 @@ func NaNPayload(f float64) (payload uint64, ok bool) {
 	return b &^ nanMask, true
 }
 
-// Nearest returns the index of the element in s
+// NearestIdx returns the index of the element in s
 // whose value is nearest to v.  If several such
 // elements exist, the lowest index is returned.
-// Panics if len(s) == 0.
-func Nearest(s []float64, v float64) int {
+// NearestIdx panics if len(s) == 0.
+func NearestIdx(s []float64, v float64) int {
+	if len(s) == 0 {
+		panic("floats: zero length slice")
+	}
+	switch {
+	case math.IsNaN(v):
+		return 0
+	case math.IsInf(v, 1):
+		return MaxIdx(s)
+	case math.IsInf(v, -1):
+		return MinIdx(s)
+	}
 	var ind int
-	dist := math.Abs(v - s[0])
+	dist := math.NaN()
 	for i, val := range s {
 		newDist := math.Abs(v - val)
-		if newDist < dist {
+		// A NaN distance will not be closer.
+		if math.IsNaN(newDist) {
+			continue
+		}
+		if newDist < dist || math.IsNaN(dist) {
 			dist = newDist
 			ind = i
 		}
 	}
 	return ind
 }
 
-// NearestWithinSpan return the index of a hypothetical vector created
+// NearestIdxForSpan return the index of a hypothetical vector created
 // by Span with length n and bounds l and u whose value is closest
-// to v. NearestWithinSpan panics if u < l. If the value is greater than u or
-// less than l, the function returns -1.
-func NearestWithinSpan(n int, l, u float64, v float64) int {
-	if u < l {
-		panic("floats: upper bound greater than lower bound")
+// to v. That is, NearestIdxForSpan(n, l, u, v) is equivalent to
+// Nearest(Span(make([]float64, n),l,u),v) without an allocation and
+// with a relaxation for the paramater n.
+// NearestIdxForSpan panics if n is less than one.
```

Comment:
This should be `n is less than two`, shouldn't it?

Reply:
Or is this what "a relaxation" means? How the hypothetical vector looks like for `n = 1`?

Reply:
Yes, I agree, this should panic for `n  < 2`, so that it matches `Span`

Reply:
OK